Nix: various improvements

faukah opened the pull request:
<!-- Thank you for making a PR! Bug fixes are always welcome, but if you're adding a new feature or changing an existing one, we'd really appreciate if you open an issue, post on the forum, or drop in on Discord -->

Cleans up the nix code in various ways, the most important ones being:
- Dropping flake-utils, thus reducing abstraction.
- Decreasing the amount of files in the `src` fileset in `atuin.nix` which prevents changes to files like README.md from changing the derivation's output, thus saving rebuils. 
- Adding a sane default formatter.

## Checks
- [X] I am happy for maintainers to push small adjustments to this PR, to speed up the review cycle
- [X] I have checked that there are no existing pull requests for the same thing

minhtrancccp commented:
would it be possible for the package to include version as well, as [`home-manager` requires that info to enable daemon](https://github.com/nix-community/home-manager/blob/0cdfcdbb525b77b951c889b6131047bc374f48fe/modules/programs/atuin.nix#L214)?

faukah committed:
atuin.nix: filter out unneeded files in src, format
Filter out files like README.md, demo.gif, flake.nix,
flake.lock to stop the atuin package from being rebuilt
as often when using the flake.

faukah committed:
flake: add default formatter

faukah committed:
flake: drop flake utils, cleanup

faukah committed:
flake.nix: format using nixfmt

faukah committed:
flake.lock: update inputs

faukah committed:
atuin.nix: add gitRev

ellie commented:
I'm sorry, we're not in a place to maintain and review this right now - happy to reopen or revisit if @arcuru or someone else who's touched our nix code in the past is willing to give it a once over

shymega commented:
@ellie I'm experienced with Nix, and contribute to Nixpkgs. If you want to ping me over the weekend, I'll take a look.

markovejnovic commented:
@shymega, would love to get your feedback! I'll open the PR right up =)

faukah commented:
I'll update it in a bit, then; it's been a while since I touched this pr :)

shymega commented:
@faukah Please ping me when you're ready - I'm off $work for the next couple of work days, so I have capacity to look at open-source projects again.

shymega commented:
@markovejnovic No worries. Yourself and @ellie are welcome to ask for my input/maintenance on Nix in the future :-)

arcuru commented:
Nix changes bring all the Nix fans out :)

I'll stay out of helping, but my drive-by suggestion is that you keep in mind the goals for the Nix stuff here:

1. Keep `nix run github:atuinsh/atuin` working, so downstream consumers (aka me) can keep running/installing from the repo.
2. A devshell with necessary tools pinned for drive-by development.
3. A dev doc, or whatever else Ellie wants, describing how to update it.
4. Don't put anything else in there.

It's tempting to move everything into Nix (I run [_everything_ inside the Nix sandbox](https://github.com/arcuru/eidetica/)) but that's explicitly not what Atuin needs.

For non-Nix people, I've found that Frontier-level LLMs are quite good at understanding and writing Nix, which should make reviewing easier.

shymega reviewed (changes requested):
I feel like this isn't necessary in this MR. It's purely stylistic, and as developers can (and do!) disagree about style, I think this commit should be dropped.

shymega reviewed (changes requested):
See: https://github.com/atuinsh/atuin/pull/2787#pullrequestreview-4835197182 regarding formatting.

Do we need this change added to the Flake? Can Atuin build without it? If so, I would argue we don't need stylistic changes.

shymega reviewed (changes requested):
See: https://github.com/atuinsh/atuin/pull/2787#pullrequestreview-4835197182 regarding formatting.

Do we need this change added to the Flake? Can Atuin build without it? If so, I would argue we don't need stylistic changes.

shymega commented on atuin.nix:
```diff
@@ -10,35 +10,43 @@
+let
+  fs = lib.fileset;
+  src = fs.difference (fs.gitTracked ./.) (
+    fs.unions [
```
You could move `fs = lib.fileset` into a let binding in `src`. No need for global scope.

shymega commented on atuin.nix:
```diff
@@ -10,35 +10,43 @@
+  gitRev,
 }:
+let
+  fs = lib.fileset;
+  src = fs.difference (fs.gitTracked ./.) (
```
Can't we just include only the workspace files, and be more explicit about what's included? I feel like filtering on Markdown, SVGs, and Git metadata, is unnecessary and a maintenance burden.

shymega commented on atuin.nix:
```diff
@@ -10,35 +10,43 @@
-  preBuild = ''
-    export LD_LIBRARY_PATH="${lib.makeLibraryPath [ openssl ]}''${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}"
-  '';
+  buildInputs = lib.optionals stdenv.isDarwin [
```
This can be `lib.optional`.

shymega commented on atuin.nix:
```diff
@@ -49,10 +57,10 @@ rustPlatform.buildRustPackage {
     homepage = "https://github.com/atuinsh/atuin";
-    license = licenses.mit;
+    license = lib.licenses.mit;
     mainProgram = "atuin";
```
Unnecessary change. The same end-result is going to happen, so no need to change this.

shymega commented on flake.nix:
```diff
@@ -12,66 +11,91 @@
       systems = [
         "x86_64-linux"
         "aarch64-linux"
         "x86_64-darwin"
```
No longer supported upstream in Nixpkgs. But maybe we should address this again later on.

shymega commented on flake.nix:
```diff
@@ -12,66 +11,91 @@
-            inherit atuin;
-            default = atuin;
-          };
+      eachSystem = lib.genAttrs systems;
+      pkgsFor = eachSystem (
+        system:
+        import nixpkgs {
+          localSystem.system = system;
+          overlays = [ self.overlays.atuin ];
+        }
+      );
```
I don't like this. I think `flake-utils` did the job just fine, but it does add to the dep graph.

At the very least, merge the two variables into something like `genAttrs systems (system: f (import nixpkgs)` and access the `system` variable from `stdenv.hostPlatform.system`. The way here for iterating over platforms is a bit strange.

shymega commented on flake.nix:
```diff
@@ -12,66 +11,91 @@
 
-              if [ -e "''${ATUIN_RECORD_STORE_PATH}" ]; then
-                echo >&2 "''${ATUIN_RECORD_STORE_PATH} already exists, you might want to double-check that"
-              fi
```
Why can't we use `pkgs.mkShell` and `inputsFrom`?

shymega commented on flake.nix:
```diff
@@ -65,10 +71,31 @@
+              gitRev = self.shortRev or "dirty";
+            };
+
+            default = self.overlays.atuin;
```
This should be addresed.

shymega commented on flake.nix:
```diff
@@ -12,66 +11,91 @@
+      });
 
-            RUST_SRC_PATH = "${pkgs.rustPlatform.rustLibSrc}";
+      devShells = lib.mapAttrs (system: pkgs: {
```
This is unusual for a Flake, and should be less complex.

shymega commented:
There's  quite a few things I've flagged, but ultimately I agree with @arcuru on this. There's changes in this PR that are a bit unusual, and I don't want to break compatibility for downstream consumers.

The KISS principle is worth bearing in mind, and we should keep the Nix Flake as simple as we can, so that Ellie and co can focus on what makes Atuin great - rather than  packaging semantics.

markovejnovic commented:
Still have no clue what I'm looking at but I want to thank @shymega for the work!!
